src/Workspaces/Core/Portable/Shared/Utilities/NameGenerator.cs

```diff
@@ -1,18 +1,59 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+#nullable enable
+
 using System;
 using System.Collections.Generic;
+using System.Collections.Immutable;
 using System.Linq;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Shared.Utilities
 {
     internal static class NameGenerator
     {
```

Comment:
this file is the meat of the change.  Nearly all functions moved to a model where they are passed a temporary array that they put their results into.  and during processing any intermediary state is handled with pooling.